2FAの登録・解除

rakushoo opened the pull request:
* Resolve #167  
* Resolve #179 

## 変更の概要

* settingから、2FA機能の有効・無効を設定できるようにした

## なぜこの変更をするのか

* 2要素認証の実装のため、登録と解除を先行して実装。

## やったこと

* [x] setting画面へ、2FA登録・解除ボタン（共通）を追加
* [x] 2FA機能の登録画面を作成(enable2fa.tsx)
* [x] 2FAの登録処理
* [x] 2FAの解除処理
* [x] 2FA登録・解除後のユーザー情報キャッシュの更新 (useMutationHas2FA.tsx)
* [x] 登録時の認証コードエラーメッセージ、解除完了のメッセージ、解除確認ダイアログを表示
* [x] トップのOAuth認証用アイコンのボタン化

## 変更内容

* setting画面のENABLE2FA（DISABLE2FAと兼用）ボタンを押下した時の動作
* 登録直後に「登録完了しました」などのアラート、 登録解除前に「解除しますか？」のダイアログと、「解除しました」のアラートなどを表示
* トップのindexで実施していたoauth認証の処理を別ファイルに移動(authenticate.tsx)
* トップの42,GoogleのOAuth用イメージがボタンのように指マークになって押せるようにした

## 影響範囲

* DBへのカラム追加などはありません。Userのhas2FA、secret2FAの値を使用するようになります。

## どうやるのか

<登録>
* setting画面で、ENABLE2FAボタンを押下
* Google AuthenticatorのインストールされたスマートフォンでQRコードを読み込み
* 6桁のコードを画面に入力し、登録ボタンを押下
* setting画面に戻り、2FAボタンがDISABLEの表示に変化している

<解除>
* setting画面で、DISABLEボタンを押下。
* 確認ダイアログでOKを押す
* setting画面で、2FAボタンがENABLEの表示に戻っている

## 課題

* ログイン時の認証（次のPRで）
* [...nextauth].ts のlintエラー（年末に相談した件）

## 備考

rakushoo committed:
fix: 2FA登録・解除をできるようにする

登録ページを追加
登録・解除時のキャッシュ更新用Mutation追加
OAuth認証の処理を別ファイルに分けた
OAuth認証の画像をボタン化

rakushoo committed:
add: 2FAの登録・解除時のメッセージとダイアログを追加

rakushoo committed:
fix: lint対策で42のOAuthではuser.emailをログイン名にするように戻した

rakushoo committed:
update: 未対応のコメントアウトコードを削除

rakushoo committed:
Merge branch 'main' of https://github.com/ryo-manba/ft_transcendence into register-2fa

takashi247 commented on frontend/pages/setting/index.tsx:
```diff
@@ -51,6 +61,8 @@ const Setting: NextPage = () => {
+  const { disableHas2FAMutation } = useMutationHas2FA();
+  const router = useRouter();
 
   if (user === undefined) return <Loading fullHeight />;
```
`router.isReady === false`という条件を追加した方がよいかと思います。
(参考: https://qiita.com/Anders/items/ad91c6752c512716f82a)

rakushoo replied:
ありがとうございます。一瞬、falseになるタイミングがあるんですね。。
ご指摘のように、67行目のifに判定を追加しました。

yuta-fujimoto commented on backend/src/auth/auth.service.ts:
```diff
@@ -192,53 +193,43 @@ export class AuthService {
     });
     if (!valid) {
-      throw new Error('hoge');
+      return 'failure';
```
戻り値boolにすればよいのではないでしょうか。

rakushoo replied:
全くその通りですね、、応答をstringで返していた関数を、booleanで返すように変更します。

yuta-fujimoto commented on frontend/hooks/useMutationHas2FA.tsx:
```diff
@@ -0,0 +1,72 @@
+    },
+  );
+
+  const disableHas2FAMutation = useMutation<
```
`enable2FA`の状態を変化させてるだけなので`disableHas2FAMutation`と`enableHas2FAMutation`は一つにまとめていいのではないでしょうか。
あと別に1234でなくて一番シンプルに空文字だとだめですか？。

rakushoo replied:
確かにそうですね。enable,disableを共通化し、changeHas2FAMutationにまとめました。

1234については、backendのAPIがauthCodeを必要としてましたが、disableの時は不要のため、idだけを受け取るように変更しました。結果的に、'1234'ではなく''で動作できるように変更しました。

yuta-fujimoto commented on frontend/pages/setting/enable2fa.tsx:
```diff
@@ -0,0 +1,151 @@
+  const { data: user } = useQueryUser();
+  const [qrCode, setQrCode] = useState('');
+  const { enableHas2FAMutation } = useMutationHas2FA();
+  const [openSnack, setOpenSnack] = useState('none');
```
開くか、開かないかなので`true`, `false`でいい気がします。

rakushoo replied:
openSnackですが、3択になっていて、デフォルトが'none'で、成功メッセージ'SUCCESS'と失敗メッセージ'ERROR'を切り分けるため、string型にしている、という事情があります。
大文字小文字などが混ざっているのでnoneをやめて、'', 'SUCCESS', 'ERROR' の三択に変更します。

yuta-fujimoto replied:
失礼しました。。理解してなかったです。修正ありがとうございます！

ryo-manba replied:
3択になっているという情報がわからないので、enumなどで定義してあげてもいいかもしれません。

yuta-fujimoto commented on frontend/pages/setting/enable2fa.tsx:
```diff
@@ -0,0 +1,151 @@
+                    clearErrors();
+                  }}
+                >
+                  REGISTERED
```
`REGISTER`ではないでしょうか？

rakushoo replied:
誤字でした。修正します。

yuta-fujimoto commented on frontend/pages/setting/index.tsx:
```diff
@@ -227,7 +269,62 @@ const Setting: NextPage = () => {
+              </DialogContent>
+              <DialogActions>
+                <Button onClick={() => handleDialogClose('agree')}>OK</Button>
+                <Button onClick={() => handleDialogClose('disagree')} autoFocus>
```
ここOKとNOの見た目を区別したほうがいいのではないでしょうか？

![image](https://user-images.githubusercontent.com/64348608/210193957-63ed806a-2636-4979-88ed-44c7647a1330.png)

こうするとかNOは赤文字にするとか
あとこれは細かいですが、`YES` `NO`のほうが返答として自然な気がします。

rakushoo replied:
CreateRoomと同様の見た目と、YES/NOに変更しました。
![スクリーンショット 2023-01-02 21 32 44](https://user-images.githubusercontent.com/2654860/210231798-b738e883-85a1-40d0-ae2a-7442ffdd5159.png)

yuta-fujimoto replied:
ありがとうございます！確認しました。

yuta-fujimoto commented on frontend/pages/setting/enable2fa.tsx:
```diff
@@ -0,0 +1,151 @@
+                  autoHideDuration={6000}
+                  onClose={handleClose}
+                >
+                  <Alert onClose={handleClose} severity="success">
```
認証に成功すると`http://localhost:3000/setting`に戻るのでこれは表示されないですね。

rakushoo replied:
画面を跨いでSnackbarを表示させるのがややこしかったので、
成功した時はsettingに戻し、表示されない指摘の箇所は削除しました。

yuta-fujimoto replied:
muiの`Dialog`とか使えるとうまくいきそうな感じもありますが、SUCCESSの表示が必要というわけでもないですね。修正ありがとうございます。

takashi247 commented on frontend/pages/setting/enable2fa.tsx:
```diff
@@ -0,0 +1,151 @@
+    }
+  };
+
+  if (qrCode == '') {
+    void onCreateComponent();
+  }
```
ここはuseEffectを使って実装した方がよいかと思いますがどうでしょうか？
(参考: https://beta.reactjs.org/learn/synchronizing-with-effects)

rakushoo replied:
ページ初期化時に実行したかったコードなので、useEffectの方が良いですね。
置き換えます。

takashi247 reviewed (commented):
PRの切り出し、ありがとうございました。2点コメントつけたので、確認お願いします。

rakushoo committed:
Merge branch 'main' of https://github.com/ryo-manba/ft_transcendence into register-2fa

rakushoo committed:
fix: 2FAの登録・解除時のAPI応答をbooleanに変更

rakushoo committed:
update: 誤字修正、snackBarのデフォルトを文字無しに変更

rakushoo committed:
update: 2FA解除時のボタン表示を変更

rakushoo committed:
fix: 2FAのmutationを共通化

rakushoo committed:
update: 実行されないコードを削除

2FA登録が成功した時は/settingに戻るため、Snackbarは表示されない。

rakushoo committed:
fix: setting画面でrouterの準備ができるまではLoadingとするよう条件を追加

rakushoo committed:
fix: 2FA登録ページの初期化処理をuseEffectで実施するようにした

yuta-fujimoto commented:
重ねて一点だけご確認お願いしたいです。(以前closeされたPRでの発言内容と趣旨は同じです。)

* 現在の状態だとdocker composeのほうで動作しない(おそらくbuildができない)感じになっていて、
`[...nextauth].ts`の27行目を
```
        token.accessToken = account.accessToken as string;
```
に変更するととりあえず動きます。これでいいのか僕もあんまりわかってないのでできたら教えてもらえると助かります。
42のAuthは通りましたが画像はまだ取れてないです。

![image](https://user-images.githubusercontent.com/64348608/210254722-136c50bc-9ade-479f-b924-461a81dfdf0d.png)

rakushoo commented:
@yuta-fujimoto 確認ありがとうございます！

> * 現在の状態だとdocker composeのほうで動作しない(おそらくbuildができない)感じになっていて、
>   `[...nextauth].ts`の27行目を
> 
> ```
>         token.accessToken = account.accessToken as string;
> ```
> 
> に変更するととりあえず動きます。これでいいのか僕もあんまりわかってないのでできたら教えてもらえると助かります。 

すみません、下記の手順を実施しましたが、状況が再現できていません。
* git clone -b register-2fa https://github.com/ryo-manba/ft_transcendence.git   で本PRのブランチを取得
* .env.localをfronten/に配置
* make コマンド実行
→ここでbuild失敗してるということでしょうか？

> 42のAuthは通りましたが画像はまだ取れてないです。  

はい、今のコードではusernameがメールアドレスになる、その状態が期待通りです。
42のログイン名と画像はAccessTokenが必要になり、そのためのInterfaceの拡張は別Issueに挙げている状態です。

yuta-fujimoto commented:
![image](https://user-images.githubusercontent.com/64348608/210298998-157e1596-c75d-40ff-9039-2d458c95c570.png)

こんな感じのエラー出てないですか？
`lint error`ですね。すみません

rakushoo committed:
fix: make実行時にdocker-composeでlint errorになるため対策を入れる

rakushoo commented:
@yuta-fujimoto ありがとうございます！
as string追加してmakeが通ることを確認したのでcommitしておきます。

yuta-fujimoto reviewed (approved):
修正ありがとうございます！approveします！！

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -192,66 +193,55 @@ export class AuthService {
 
-    return 'ok';
+    return true;
   }
```
updateは例外処理があったほうが良さそうです！（trycatchで囲む）
https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#return-type-4

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -272,18 +262,17 @@ export class AuthService {
+  async disable2FA(id: string): Promise<boolean> {
     const user_db = await this.prisma.user.update({
       where: {
-        id: Number(data.userId),
+        id: Number(id),
       },
       data: {
         has2FA: false,
         secret2FA: '',
       },
     });
 
-    // TODO: 戻り値がstringで良いか、frontendの実装時に再検討予定
-    return 'disabled: true';
+    return true;
```
戻り値がbooleanに対して、true以外変えるパターンがないのは不自然な気がします。
updateに失敗した場合に、falseを返してあげるといいと思います！

rakushoo replied:
ありがとうございます。上の指摘と合わせて例外でfalseを返すようにしました。

ryo-manba commented on frontend/hooks/useMutationHas2FA.tsx:
```diff
@@ -0,0 +1,55 @@
+          void router.push('/setting'); //登録が成功したらsettingに戻る
+        } else {
+          // codeの数字が間違った場合
+          throw new Error('Register ERROR Please Retry');
+        }
+      },
+      onError: (err: AxiosError) => {
+        // codeに数字以外が含まれる場合
+        console.log(err);
+        throw err;
+      },
```
49行目の `throw err` を `throw new Error(原因)` にしてあげたほうが統一感あっていいのかなと思いました！

ryo-manba commented on frontend/pages/authenticate.tsx:
```diff
@@ -0,0 +1,53 @@
+            } else if (session.user.email) {
+              // TODO: 42のアカウントでは、access_tokenからログインIDと画像を取得する必要がある。別PRで対応
+              login = session.user.email;
+              image_url = '';
```
キャメルケースにしたほうが統一感あっていいですね

ryo-manba commented on frontend/pages/authenticate.tsx:
```diff
@@ -0,0 +1,53 @@
+  const router = useRouter();
+  const { data: session, status } = useSession();
+  useEffect(() => {
+    const f = async () => {
```
fという関数名だと何をしているか分かりづらいので名前つけていただきたいです！

takashi247 commented on frontend/hooks/useMutationHas2FA.tsx:
```diff
@@ -0,0 +1,55 @@
+        const { data } = await axios.patch<boolean>(
+          `${process.env.NEXT_PUBLIC_API_URL as string}/auth/send2facode`,
+          {
+            userId: String(userId),
```
ここのuserIdですが、バックエンドに送られたあと、またNumberを使ってnumber型に変換し直しているかと思うんですが、そもそも、ここで送るときにnumberのままで送ってしまえばいいのではないかと思いました。ここで一回string型に直してる理由って何かありますでしょうか？

rakushoo replied:
ご指摘ありがとうございます！故意ではなくURIのクエリが文字列だったため、それに引きずられてStringになっていました。userIdはDB上もnumberで扱っているため、型を揃えて対応しました。

takashi247 reviewed (commented):
修正ありがとうございました。申し訳ないんですが、もう一点、追加でコメントをつけたので、確認お願いします。

rakushoo committed:
fix: 2FAのAPIでDBの例外時はfalseを返すようにした

rakushoo committed:
update: レビュー指摘対応

例外の返し方を修正
変数名称をcamelCaseにした。

rakushoo committed:
update: OAuth認証時の関数名称を変更

rakushoo committed:
fix: APIのidをnumberに変更

URIで受け取るときはstringになってるが、serviceではnumberで扱うように統一
authCodeはfrontendでは入力値に制限をつけてnumber、backendではライブラリの引数の都合上stringで扱うこととした。

ryo-manba commented on frontend/pages/setting/enable2fa.tsx:
```diff
@@ -0,0 +1,142 @@
+          }
+        } catch (e) {
+          console.log(e);
+        }
```
QRコードの取得に失敗したら、それ以降の処理って行えないですよね？？
例外補足したら、`throw new Error` でもいいのかなと思いました！

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -192,66 +193,59 @@ export class AuthService {
+    }
+    //2FAの登録が完了したら、2FA機能をオンにして登録
+    try {
+      const user_db = await this.prisma.user.update({
```
ここもキャメルケースでお願いします！

rakushoo replied:
updateの戻り値は参照していなかったので削除で対応しました。

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -272,18 +266,21 @@ export class AuthService {
-    });
+  async disable2FA(id: number): Promise<boolean> {
+    try {
+      const user_db = await this.prisma.user.update({
```
ここもキャメルケースでお願いします！

rakushoo replied:
同じくこちらも削除で対応しました。

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -192,66 +193,59 @@ export class AuthService {
+    // 一時的にシークレットを保存
+    this.preAuthSecrets.set(userId, secretBase32);
 
     return qr_code;
```
ここもキャメルケースでお願いします！

ryo-manba commented on frontend/pages/authenticate.tsx:
```diff
@@ -0,0 +1,53 @@
+              login = session.user.email;
+              imageUrl = '';
+            }
+            const url_oauth = `${process.env.NEXT_PUBLIC_API_URL}/auth/oauth-login`;
```
ここもキャメルケースでお願いします！

ryo-manba commented on frontend/pages/setting/index.tsx:
```diff
@@ -227,7 +272,71 @@ const Setting: NextPage = () => {
+          </Grid>
+        </Grid>
       </form>
+      {/* <Grid container alignItems="center" justifyContent="center" spacing={2}>
+        <Grid item>
+          <Enable2FA></Enable2FA>
+        </Grid>
+      </Grid> */}
```
コメント部分使用予定内容であれば、削除お願いします！

rakushoo committed:
fix: QRコードの取得に失敗したら例外出すように変更

rakushoo committed:
update: 変数名をcamelCaseに統一

コメントアウトのコードを削除

takashi247 commented on frontend/types/setting.ts:
```diff
@@ -1,3 +1,7 @@
 };
+
+export type TwoAuthForm = {
+  authCode: number;
```
authCodeって0からスタートする場合とかもあると思うので、numberではなくstring型で管理したほうがいいのではないかと思うのですがどうでしょうか？

rakushoo replied:
まったくもってそうでした！6桁のコードで012345の場合もあるので、numberだとまずいですね。
stringに変更しました。

takashi247 commented on backend/src/auth/auth.controller.ts:
```diff
@@ -124,13 +122,11 @@ export class AuthController {
-  disable2FA(@Body() data: Validate2FACodeDto): Promise<string> {
-    return this.authService.disable2FA(data);
+  @Patch('disable2fa/:id')
+  disable2FA(@Param('id') id: string): Promise<boolean> {
```
ここも上と同じで、`@Param('id', ParseIntPipe) id: number`への修正をお願いします。

takashi247 commented on backend/src/auth/auth.controller.ts:
```diff
@@ -98,24 +99,21 @@ export class AuthController {
 
   @Get('has2fa')
-  has2FA(@Param('id') id: string): Promise<string> {
+  has2FA(@Param('id') id: string): Promise<boolean> {
```
下のdiable2FAの部分の修正を見て気づいたのですが、ここの部分は`@Param('id', ParseIntPipe) id: number`という形にしてもらうと、server-side validationとnumber型への変換を同時に行ってくれるので、そのように修正してもらえますでしょうか？（109行目のNumber()というのも不要になります。）
課題文の中でserver-side validationが求められているので、追加しておきたいです。

rakushoo replied:
ありがとうございます。やはりAPI経由ではStringになってしまうんですよね。
変換（とバリデーション）やってくれるとのことでidを取得しているAPIにParseIntPipeを追加しました。

takashi247 replied:
axiosに`params`として渡した場合にはURLの後ろにクエリの形で変数が追加されるので、全てString型になってしまいますが、リクエストのBodyとして渡した場合にはString型にはならないかと思います。（例えば`send2FACode`の方の`dto.userId`はフロントエンドからバックエンドにnumber型のまま送れているかと思います。）
今回は`params`として引数を渡しているためにフロントエンドで何の型の変数を渡していたとしてもString型に勝手に変換されてしまう（逆に言うと、自分でフロントエンドでString型に変換する必要はない）、ということかと思います。

takashi247 reviewed (commented):
修正ありがとうございます！追加コメントの確認をお願いします。

ryo-manba reviewed (approved):
度々修正いただきありがとうございます！！自分が気になった点は大丈夫だと思います！
残りはnishinaさんのコメントの対応お願いします！

rakushoo committed:
fix: authCodeをstring型に変更

APIでstringで受け取ったものをnumber変換しバリデーションを行う

rakushoo committed:
fix: authCodeの型変更の対応漏れを修正

takashi247 reviewed (approved):
早速の修正、ありがとうございました！マージします！
